lib/stdlib/src/binary.erl

```diff
@@ -365,61 +365,124 @@ get_opts_replace(_,_) ->
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% Hex encoding functions
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+-define(HEX(X), (hex(X)):16).
 -spec encode_hex(Bin) -> Bin2 when
       Bin :: binary(),
-      Bin2 :: binary().
-encode_hex(Bin) when is_binary(Bin) ->
-    encode_hex(Bin, <<>>);
+      Bin2 :: <<_:_*16>>.
+encode_hex(Data) when byte_size(Data) band 7 =:= 0 ->
+    << <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E),?HEX(F),?HEX(G),?HEX(H)>> || <<A,B,C,D,E,F,G,H>> <= Data >>;
+encode_hex(Data) when byte_size(Data) rem 7 =:= 0 ->
+    << <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E),?HEX(F),?HEX(G)>> || <<A,B,C,D,E,F,G>> <= Data >>;
+encode_hex(Data) when byte_size(Data) rem 6 =:= 0 ->
+    << <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E),?HEX(F)>> || <<A,B,C,D,E,F>> <= Data >>;
+encode_hex(Data) when byte_size(Data) rem 5 =:= 0 ->
+    << <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E)>> || <<A,B,C,D,E>> <= Data >>;
+encode_hex(Data) when byte_size(Data) rem 4 =:= 0 ->
+    << <<?HEX(A),?HEX(B),?HEX(C),?HEX(D)>> || <<A,B,C,D>> <= Data >>;
+encode_hex(Data) when byte_size(Data) rem 3 =:= 0 ->
+    << <<?HEX(A),?HEX(B),?HEX(C)>> || <<A,B,C>> <= Data >>;
+encode_hex(Data) when byte_size(Data) rem 2 =:= 0 ->
+    << <<?HEX(A),?HEX(B)>> || <<A,B>> <= Data >>;
+encode_hex(Data) when is_binary(Data) ->
+    << <<?HEX(N)>> || <<N>> <= Data >>;
```

Comment:
Have you tested it with

```suggestion
encode_hex(Data) when byte_size(Data) > 7 ->
    Size = byte_size(Data),
    Mul8 = 8 * (Size div 8),
    <<Prefix:Mul8/binary, Suffix/binary>>,
    PrefixEncoded =<< <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E),?HEX(F),?HEX(G),?HEX(H)>> || <<A,B,C,D,E,F,G,H>> <= Prefix >>,
    SuffixEncoded = encode_hex(Suffix),
    <<PrefixEncoded/binary, SuffixEncoded/binary>>;
encode_hex(<<A,B,C,D,E,F,G>>) ->
    <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E),?HEX(F),?HEX(G)>>;
encode_hex(Data) when byte_size(Data) ->
    <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E),?HEX(F)>>;
encode_hex(<<A,B,C,D,E>>) ->
    <<?HEX(A),?HEX(B),?HEX(C),?HEX(D),?HEX(E)>>;
encode_hex(<<A,B,C,D,E>>) ->
    <<?HEX(A),?HEX(B),?HEX(C),?HEX(D)>>;
encode_hex(<<A,B,C>>) ->
    <<?HEX(A),?HEX(B),?HEX(C)>>;
encode_hex(<<A,B>>) ->
    <<?HEX(A),?HEX(B)>>;
encode_hex(<<A>>) ->
    <<?HEX(N)>>;
encode_hex(<<>>) ->
    <<>>;
```

Reply:
No taking into account that the snippet doesn't compile (but it was obvious how to fix it), this wouldn't bring better results, and it doesn't make the code size (.beam file) smaller either:

```
##### With input 1. Tiny #####
Name                       ips        average  deviation         median         99th %
encode optimised      935.95 K        1.07 μs  ±3968.58%        0.75 μs        2.21 μs
encode suggested      641.82 K        1.56 μs  ±3046.43%        1.11 μs        2.93 μs

Comparison:
encode optimised      935.95 K
encode suggested      641.82 K - 1.46x slower +0.49 μs

Memory usage statistics:

Name                Memory usage
encode optimised           128 B
encode suggested           248 B - 1.94x memory usage +120 B

**All measurements for memory usage were the same**

##### With input 4. Medium Big #####
Name                       ips        average  deviation         median         99th %
encode optimised       38.58 K       25.92 μs   ±182.31%       24.46 μs       40.97 μs
encode suggested       34.58 K       28.92 μs   ±291.90%       27.05 μs       45.45 μs

Comparison:
encode optimised       38.58 K
encode suggested       34.58 K - 1.12x slower +3.00 μs

Memory usage statistics:

Name                Memory usage
encode optimised           128 B
encode suggested           264 B - 2.06x memory usage +136 B

**All measurements for memory usage were the same**

##### With input 6. Large #####
Name                       ips        average  deviation         median         99th %
encode optimised        321.56        3.11 ms     ±6.02%        3.10 ms        3.30 ms
encode suggested        312.56        3.20 ms     ±6.48%        3.20 ms        3.41 ms

Comparison:
encode optimised        321.56
encode suggested        312.56 - 1.03x slower +0.0895 ms

Memory usage statistics:

Name                Memory usage
encode optimised           128 B
encode suggested           264 B - 2.06x memory usage +136 B

**All measurements for memory usage were the same**
```